File: pawn.cs
/**
 * @file pawn.cpp
 * @author rtorigoe
 * @version 0.1
 * @date 2023-09-16
 */

/**
 * CLASS INVARIANT
 * 
 * Exceptions: CLIENT MUST HANDLE ALL EXCEPTIONS
 * 
 * 
 */

namespace chess {
    class piece
    {
        protected const int BOARD_WIDTH = 8;
        protected const int BOARD_LENGTH = 8;
        protected const int POINT_VAL = 1;
        protected const bool COLOR; // true for white, false for black
        protected int X_COORD;
        protected int Y_COORD:
        
    }

    /**
     * Pre-Conditions:
     *     - x is within the valid range of values (0 - BOARD_WIDTH - 1)
     *     - y is within the valid range of values (0 - BOARD_LENGTH - 1)
     *
     * Post-Conditions:
     *     - object's x and y values will change to the parameter values  
    **/

    protected bool makeMove(int x, int y) {
        if ((x < 0) || (x > BOARD_WIDTH)) {
            throw new ArgumentOutOfRangeException("coordinate out of bounds");
        }
        if ((y < 0) || (y > BOARD_HEIGHT)) {
            throw new ArgumentOutOfRangeException("coordinate out of bounds");
        }
        if (this.X_COORD != 0) {
            return false;
        }
        if (this.COLOR == true) {
            if (this.Y_COORD - y != 1) {
                return false;
            }
        }
        if (this.COLOR == false) {
            if (this.Y_COORD - y != -1) {
                return false;
            }
        }
        
        this.x = x;
        this.y = y;
        return true;
    }

    /**
     * IMPLEMENTATION INVARIANT
     *
     * Members: 
     *     - BOARD_LENGTH, BOARD_WIDTH, COLOR, AND POINT_VAL are kept constant to refrain from changes
     *     - COLOR is stored as a boolean to represent two possibilities of color (black or white)
     * 
     * Methods:
     *     - makeMove:
     *         - ArgumentOutOfRange exception thrown when the given coordinate is off the board
     *         - Checks the coordinates to see if they're valid coordinates, but invalid to move to
     *         - true is returned upon a correct call to the method, false returned otherwise
    **/
}

File: pawn.java
/**
 * @file pawn.cs
 * @author rtorigoe
 * @version 0.1
 * @date 2023-09-16
 */

/**
 * CLASS INVARIANT
 * 
 * Exceptions: CLIENT MUST HANDLE ALL EXCEPTIONS
 * 
 */

import java.lang.Math;

public class pawn extends piece
{
    private boolean PROMOTABLE;
    private boolean UNMOVED;

    /**
     * @brief ctor
     * 
     * @param x - int, valid values of 0-7
     * @param y - int, valid values of 0-7
     * @param color - bool, valid values of true or false
     * @return public pawn obj
     */

    public pawn(int x, int y, boolean color) {
        super(x, y, color);
        this.PROMOTABLE = false;
        this.UNMOVED = true;
    }

    /**
     * Pre-Conditions:
     *     - x is within the valid range of values (0 - BOARD_WIDTH - 1)
     *     - y is within the valid range of values (0 - BOARD_LENGTH - 1)
     * 
     * Post-Conditions:
     *     - object's x and y values will change to the desired values
     */

    @Override public boolean move(int x, int y) {
        try {
            if ((X_COORD - x != 0) || Math.abs(Y_COORD - y) > 2) {
                throw new Exception("invalid coordinate to move to");
            }
            if (Math.abs(Y_COORD - y) == 2 && (UNMOVED == false)) {
                throw new Exception("pawn cannot move 2 spaces, has moved already");
            }
            Y_COORD = y;
            UNMOVED = false;
            return true;
        }
        catch (Exception e) {
            e.getMessage();
        }
        return false;
    }
}

/**
 * IMPLEMENTATION INVARIANT
 * 
 * Members:
 *     - PROMOTABLE and UNMOVED are stored as booleans to represent dicotomy (either true or false)
 * 
 * Constructor:
 *     - Sets PROMOTABLE to false by default
 *     - Sets UNMOVED to true by default
 * 
 * Methods:
 *     - move():
 *         - Exceptions thrown if the pawn cannot move to the desired space
 *         - Otherwise sets the y coordinate to the desired location and sets UNMOVED to false
 */